### PR TITLE
Convert Imports to Indirect Calls for Dynamic Linking

### PR DESCRIPTION
**Pull request by awtcode**
Inter-module calls in dynamic linking currently go thru JS stubs. Convert the calls to Wasm Indirect calls for the following reasons:

1. Faster performance
2. Skip legalization
3. Remove JS stubs (To be done in a later PR)

The Emscripten portion of the work is done in this PR https://github.com/emscripten-core/emscripten/pull/10003

### CODE REVIEW

**Comment by awtcode** [comment]
@sbc100 , this is the corresponding work in Binaryen.

File: src/asm2wasm.h [hunk 1]
```diff
@@ -1696,6 +1696,7 @@ void Asm2WasmBuilder::processAsm(Ref ast) {
     passRunner.add("merge-blocks");
     passRunner.add("optimize-instructions");
     passRunner.add("post-emscripten");
+    passRunner.add("ImportsToIndirectCalls");
```
Comment (sbc100) [inline src/asm2wasm.h]: You probably don't want to add this here.

File: src/tools/wasm-emscripten-finalize.cpp [hunk 2]
```diff
@@ -254,6 +255,15 @@ int main(int argc, const char* argv[]) {
     }
   }
 
+  // Convert the imports to indirects before the Legalization pass to 
+  // avoid unnecessary legalization
+  if (!options.passOptions.arguments["library-file"].empty()) {
```
Comment (sbc100) [inline src/tools/wasm-emscripten-finalize.cpp]: Can you give this argument a better name too?   How about "js-symbols"?

**Comment by awtcode** [comment]
> Is this showing improvements for you?

I have updated the Emscripten PR with the benchmark results. On Chrome, we see a 13x improvement while on FF, we see a 7x improvement. 

> 
> Are you running it both on the side module and the main module?
> 
We only need to run this on the main module because the previous PR https://github.com/emscripten-core/emscripten/pull/9917 already ensures that we directly call an exported Wasm function rather than the JS stub. This is already faster than what this PR provides, which is to make an indirect call to an exported Wasm function.

> In general I think this looks a lot like what we talked about. I'd like to see a test or two added in the `test/lld` directory to see the effect of this on the code.
> 

Sure, I can always add my benchmark as a testcase to the repo.

> Thanks for your work on the Kevin. I don't think we are ready to land this yet but maybe with a few iterations and a little more investigation on our side. @kripken suggested we come up with a design doc for example.

Sounds good, can you point me to an example design doc in case you have a specific format in mind?

File: src/passes/ImportsToIndirectCalls.cpp [hunk 3]
```diff
@@ -0,0 +1,213 @@
...
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+//
+// Turn indirect calls into direct calls. This will improve the runtime
```
[the remark this answers is not in the document]
Comment (awtcode) [reply src/passes/ImportsToIndirectCalls.cpp]: Oops sorry, will update the wrong comment.

File: src/tools/wasm-emscripten-finalize.cpp [hunk 4]
```diff
@@ -32,6 +32,7 @@
 #include "wasm-io.h"
 #include "wasm-printing.h"
 #include "wasm-validator.h"
+#include "C:\git\binaryen\binaryen\src\passes/passes.h"
```
[the remark this answers is not in the document]
Comment (awtcode) [reply src/tools/wasm-emscripten-finalize.cpp]: Oops, I forgot about this. Will switch to relative paths.

File: src/tools/wasm-emscripten-finalize.cpp [hunk 5]
```diff
@@ -254,6 +255,15 @@ int main(int argc, const char* argv[]) {
     }
   }
 
+  // Convert the imports to indirects before the Legalization pass to
```
[the remark this answers is not in the document]
Comment (awtcode) [reply src/tools/wasm-emscripten-finalize.cpp]: Both of them slow down the performance.

File: src/asm2wasm.h (hunk 1 above)
Comment (awtcode) [reply src/asm2wasm.h]: Ok, can you suggest a better place to add this?

File: src/passes/ImportsToIndirectCalls.cpp [hunk 6]
```diff
@@ -0,0 +1,213 @@
...
+
+  void visitCall(Call* curr) {
+    auto name = curr->target;
+
+    if (!name.isNull()) {
+      auto module = getModule();
+      auto* func = module->getFunction(name);
+      
+      if (func) {
+        if (func->imported()) {
```
[the remark this answers is not in the document]
Comment (awtcode) [reply src/passes/ImportsToIndirectCalls.cpp]: ok

File: src/passes/ImportsToIndirectCalls.cpp [hunk 7]
```diff
@@ -0,0 +1,213 @@
...
+    import->type = functionType->name;
+    FunctionTypeUtils::fillFunction(import, functionType);
+    module->addFunction(import);
+    return import;
+  }
+
+  void visitCall(Call* curr) {
+    auto name = curr->target;
+
+    if (!name.isNull()) {
```
[the remark this answers is not in the document]
Comment (awtcode) [reply src/passes/ImportsToIndirectCalls.cpp]: ok

File: src/tools/wasm-emscripten-finalize.cpp (hunk 2 above)
Comment (awtcode) [reply src/tools/wasm-emscripten-finalize.cpp]: Sure, sounds good.

File: src/asm2wasm.h (hunk 1 above)
Comment (awtcode) [reply src/asm2wasm.h]: This was a mistake. Will remove it.

File: src/passes/ImportsToIndirectCalls.cpp [hunk 8]
```diff
@@ -0,0 +1,213 @@
...
+
+    myfile.seekg(0, std::ios::end);
+    str.reserve(myfile.tellg());
+    myfile.seekg(0, std::ios::beg);
+
+    str.assign((std::istreambuf_iterator<char>(myfile)),
+               std::istreambuf_iterator<char>());
+
+    // index.wasm.libfile will have the following format:
+    // ['_glClearStencil', '_glUniformMatrix2fv', '_eglGetCurrentSurface']
```
[the remark this answers is not in the document]
Comment (awtcode) [reply src/passes/ImportsToIndirectCalls.cpp]: How about 'index.js.symbols' since the file contains JS symbols?

**Comment by kripken** [comment]
Thinking about this some more, I still wish we could simplify everything - make all calls indirect, for example - but I agree that adds some risk and downsides. So I am mostly ok with this approach, if @sbc100 you think this is maintainable and does not add any problems for migrating to a long-term better model for dynamic linking.

I'd suggest restructuring this as follows: Not run the pass from finalize, and instead, it will be run from emscripten. A pass can take arguments (see e.g. `ExtractFunction` and `Asyncify` for examples), so it could have an `ignoreList` param that has imports to ignore (which would be JS imports).

Then on the Emscripten side, we can run this at the appropriate time. The main `BINARYEN_PASSES` infrastructure is what drives the asyncify pass, for example, and this might make sense there. Note that this happens after the JS compiler runs, so we don't need to run it twice - the single invocation can provide us the list of JS imports, and we can stash it in an internal setting.

**Comment by awtcode** [comment]
> I'd suggest restructuring this as follows: Not run the pass from finalize, and instead, it will be run from emscripten. A pass can take arguments (see e.g. `ExtractFunction` and `Asyncify` for examples), so it could have an `ignoreList` param that has imports to ignore (which would be JS imports).

@kripken , running this pass in finalize has the added benefit of doing the imports to indirects conversion before legalization and that will help in both reducing the size of the binary as well as runtime performance. Unless we move the legalization pass out of finalize as well?

File: src/passes/pass.cpp [hunk 9]
```diff
@@ -342,6 +342,9 @@ void PassRegistry::registerPasses() {
                "removes local.tees, replacing them with sets and gets",
                createUnteePass);
   registerPass("vacuum", "removes obviously unneeded code", createVacuumPass);
+  registerPass("ImportsToIndirectCalls",
+               "turns imports into indirect calls",
```
[the remark this answers is not in the document]
Comment (awtcode) [reply src/passes/pass.cpp]: Updated it to "convert calls to imported functions into indirect calls for the main module"

File: src/passes/ImportsToIndirectCalls.cpp [hunk 10]
```diff
@@ -0,0 +1,213 @@
...
+  void run(PassRunner* runner, Module* module) override {
+    Name name = runner->options.getArgument(
+      "library-file",
+      "ImportsToIndirectCalls usage:  wasm-emscripten-finalize --pass-arg=library-file@FILE_NAME");
+
+    std::set<std::string> libraryFunctions = getLibraryFunctions(name);
+
+    if (!module->table.exists) {
+      Fatal() << "Table does not exist in module!!!";
+      return;
```
[the remark this answers is not in the document]
Comment (awtcode) [reply src/passes/ImportsToIndirectCalls.cpp]: Ok sure.

**Comment by kripken** [comment]
@awtcode Interesting question about the legalize pass. Yes, it seems like this should happen before legalize. In fact it seems like moving legalize to the very end is the most efficient thing in general. I think that's worth trying, although there may be issues I don't remember right now. @sbc100 what do you think?

**Comment by kripken** [comment]
I did some investigating of moving legalization all the way to the end of the emcc.py pipeline. The first serious issue I hit was that if we legalize in finalize, which is early, then we find out if we need `getTempRet0/setTempRet0` at that time, and can tell the JS compiler to emit it or not. If we legalize late, we'd be forced to always emit those, and rely on optimizations to remove them like metadce. That feels like it might be annoying enough that it's not worth the change, curious what others think.

**Comment by sbc100** [comment]
If we always emit getTempRet0/setTempRet0 can't meta-dce then remove them?

**Comment by kripken** [comment]
Yes, metadce would, so this is not going to increase fully optimized builds. It's still annoying though I think.

**Comment by awtcode** [comment]
Not required anymore.

**Comment by awtcode** [comment]
Not required anymore.

**Pull request closed without merging**